### ISSUE TITLE
Autoscaler couldn't be configured if machines are unheatlhy

### ISSUE THREAD

**Issue by himanshu-kun** [gardener/gardener#7057]
**How to categorize this issue?**

/area auto-scaling
/area performance
/area robustness
/kind bug

**What happened**:

Customers using gardener shoot clusters have faced issues where they run out of quota for a particular worker group and to avoid being stuck with machines trying to come-up , they increase the min size of another similar worker group which has quota. Sometimes they even introduce the new worker pool ,but this setting doesn't reach till autoscaler because autoscaler is updated only when all worker pools are healthy. 

**What you expected to happen**:
Autoscaler flags to update even when worker pools are unhealthy

**How to reproduce it (as minimally and precisely as possible)**:
N.A.

**Anything else we need to know?**:

To resolve the issue currently , operator has to manually update the autoscaler deployment flags , but this solution also faces problem if the update is not done for `worker` resource , as the worker controller would then keep on updating the machineDeployment with the old `min`.

**Environment**:

- Gardener version: 1.59.0
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others:

**Comment by himanshu-kun** [comment]
cc @rishabh-11

**Comment by rishabh-11** [comment]
In CA, we have three different scale-down logics:-
1. Removal of long unregistered nodes - CA will scale down the corresponding node group if a node is not registered for `max-node-provision-time`. The logic for this can be checked out [here](https://github.com/gardener/autoscaler/blob/4d7787f19559533c5f2e9a0aef710b9000e84523/cluster-autoscaler/core/static_autoscaler.go#L350).
2. Removal of unneeded nodes - If a node has low resource utilization for `scale-down-unneeded-time`, CA will scale down the corresponding node group.
3. Fix Node Group Size logic - If the number of machines is below the acceptable range, then CA scales down the corresponding node group to the current number of machines in the node group. The logic for this can be checked out [here](https://github.com/gardener/autoscaler/blob/4d7787f19559533c5f2e9a0aef710b9000e84523/cluster-autoscaler/core/static_autoscaler.go#L381). The code to calculate the acceptable range for a node group can be checked [here](https://github.com/gardener/autoscaler/blob/4d7787f19559533c5f2e9a0aef710b9000e84523/cluster-autoscaler/clusterstate/clusterstate.go#L317).

I experimented to check how scale-down behaves in specific scenarios:-
1. The number of machines is lower than the min limit or higher than the max limit specified in the CA flags for the node group:- The CA does not enforce the min/max limits for a node group. So, it will not scale up the node group to reach the min limit of the node group, and it will not scale down the node group to reach the max limit of the node group. There is a question in the FAQ of CA for this. It can be checked out [here](https://github.com/gardener/autoscaler/blob/machine-controller-manager-provider/cluster-autoscaler/FAQ.md#my-cluster-is-below-minimum--above-maximum-number-of-nodes-but-ca-did-not-fix-that-why). 
2. E.g.:- There is a node group with `3 `nodes where `1` is `Ready` and `2` are`LongUnregistered` (unregistered for `max-node-provision-time`). In the CA flags, this node group's min/max limit is `3:5`(`3`-> min and `5`-> max). In this case, CA will not scale down the node group. This is because the `long unregistered` scale-down logic does not scale down below the min limit of the node group. The number of registered nodes is within the acceptable range calculated [here](https://github.com/gardener/autoscaler/blob/4d7787f19559533c5f2e9a0aef710b9000e84523/cluster-autoscaler/clusterstate/clusterstate.go#L488)(acceptable range will be `[1,3]` in this case), so the `fixNodeGroupSize` logic also won't kick in.
3. We have a node group with some registered nodes that are **NOT** in a `Ready` state for `max-node-provision-time`. In this case, the CA will scale down the node group using the `fixNodeGroupSize` logic, causing the removal of such machines. This was noticed in [this](https://github.tools.sap/kubernetes-live/issues-live/issues/2309) issue.

For the Scale-Up of node groups, there should be unschedulable pods, and no upcoming nodes or upcoming nodes should be in a node group for which CA has a backoff for scale-up selection.

After this experiment, we think we should deploy CA even if we have an `Unhealthy` worker pool. We also plan to remove the `fixNodeGroupSize` logic along with it.

WDYT @vlerenc, @rfranzke ??

**Comment by vlerenc** [comment]
Unrelated, we discussed a similar matter: In the case of a zone outage, our reconciliation flow logic is too rigid. Unfortunately, it is important to be rigid in some scenarios like upgrading your Kubernetes cluster (so that you not speak unexpectedly to an old incompatible API server) or credentials rotation (so that you do not get locked out with step 2). However, in the case of a zone outage (or similar scenarios), the/any readiness checks block the flow from reaching a step/the step the end user/operator wanted to reach like adding a new worker pool, increasing the size, reconfiguring CA to mitigate a zone outage or ongoing resource crunch. As we couldn't see a solution with acceptable effort, we settled on https://github.com/gardener/gardener/issues/7224. With this (in development by @timuthy), the operator can help herself in cases they need to "push changes through".

**Comment by vlerenc** [comment]
Going back to your subject matter, the above "way out" should not hold us back from individually improving/fine-tuning flow steps (and making them more independent and resilient). If you believe, that CA can be deployed/updated w/o "healthy" worker pools and this won't break other scenarios (like the ones mentioned above), we should absolutely do that. If however, a "healthy" worker pool carries meaning beyond "understanding the state of a worker pool" and flow steps (dependencies) are based on that, one must be careful not to break scenarios that depend on this "healthiness" meaning.

I guess, it is about the following flow step: https://github.com/gardener/gardener/blob/master/pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go#L570-L574 ? Unfortunately, that seems to wait for the worker pools to be reconciled... and when I dig deeper, the cloud config to have been updated and acknowledged. I am not even sure whether @timuthy's work would include it and it sounds like a problematic use case (and not only about the CA's views on the healthiness of the worker pools), does it not?

So, we must be careful. We want operators to successfully update the configuration, especially when being blocked by the very same thing they want to repair, but we must also be careful, not to introduce risks and (timing-sensitive) uncertainty in the flow, because certain assumptions won't hold true anymore (and if the system is too slow, surprising bugs will pop up).

**Comment by himanshu-kun** [comment]
Thanks @vlerenc for your inputs. I think if no-one has any other concerns , we'll proceed with our work on this, as we have proposed, and raise a PR to remove `deployClusterAutoscaler` dependency on `waitUntilWorkerReady`.

**Comment by dimakow** [comment]
@himanshu-kun given your [comment](https://github.tools.sap/kubernetes-live/issues-live/issues/2500#issuecomment-2010704) I would like to ask if you already have a rough ETA, at least for the removal ([described here above](https://github.com/gardener/gardener/issues/7057#issuecomment-1379989739)) of `deployClusterAutoscaler` on `waitUntilWorkerReady` ?

If I sum up correctly, the proposed removal would have helped to prevent https://github.tools.sap/kubernetes-live/issues-live/issues/2500#issuecomment-2010704.

**Comment by himanshu-kun** [comment]
The work is ongoing , it could take upto a week for it to finish. @rishabh-11 would raise a PR soon.

**Comment by sunir1** [comment]
@rishabh-11 can you please update if a PR was already performed?

**Comment by rishabh-11** [comment]
@sunir1 I still need to raise a PR. I am currently testing the change I made and will raise a PR when everything works as expected.

**Comment by vlerenc** [comment]
Hi colleagues, some stakeholders waiting for this feature were asking whether you have some update?

Will the discussed solution work or did you hit a wall? Just asking...

**Comment by rishabh-11** [comment]
Hi @vlerenc, we had kept this on hold because of the backlog grooming of MCM/CA. The solution I was working on before was not working when I tested it. I have to figure out a new solution for this. I will resume working on a solution for this and update it here.

**Comment by elankath** [comment]
### Background of current problem.

The current shoot reconciliation flow goes like this:

#### Shoot Reconciler
Inside
[pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go](https://github.com/gardener/gardener/blob/v1.65.2/pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go#L659)

➡ `shoot.Reconciler.Reconcile` 
➡ `shoot.Reconciler.runReconcileShootFlow`
➡ `g = flow.NewGraph(...)` # initialize shot reconciliation task flow graph
➡ ... prev flow tasks ... 
➡ `deployWorker` # flow task
➡ `waitTillWorkerReady` # flow task
➡ `deployClusterAutoscaler` # flow task

#### deployClusterAutoscaler
[deployClusterAutoscaler](https://github.com/gardener/gardener/blob/v1.65.2/pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go#L659) is a flow task that encapsulates
[botanist.DeployClusterAuthoscaler](https://github.com/gardener/gardener/blob/v1.65.2/pkg/operation/botanist/clusterautoscaler.go#L44) which launches the cluster autoscaler after grabbing the slice of wanted `MachineDeployments` from the worker extension

```go		
b.Shoot.Components.ControlPlane.ClusterAutoscaler.SetMachineDeployments(
        b.Shoot.Components.Extensions.Worker.MachineDeployments());
 b.Shoot.Components.ControlPlane.ClusterAutoscaler.Deploy(ctx);
```
The problem is that the machine deployments slice in the worker extension are only populated after the worker extension is ready.

#### waitUntilWorkerReady

`waitUntilWorkerReady` delegates to

➡   [botanist.Shoot.Components.Extensions.Worker.Wait](https://github.com/gardener/gardener/blob/v1.65.2/pkg/operation/botanist/component/extensions/worker/worker.go#L286)
which waits until the [Worker extension](https://github.com/gardener/gardener/blob/v1.65.2/pkg/apis/extensions/v1alpha1/types_worker.go#L41)  is ready
➡  `extensions.WaitUntilExtensionObjectReady(...,worker, postReadyFunction)`
➡  The `postReadyFunction` is a lambda that grabs the machine deployments slice from the worker extension and saves it in the controller struct, but only **after** the worker extension is ready.
```go
func() error {
  w.machineDeployments = w.worker.Stsatus.MachineDeployments
  return nil
},
```
An extension is considered ready using [health.CheckExtensionObject](https://github.com/gardener/gardener/blob/v1.65.2/pkg/utils/kubernetes/health/extensions.go#L35) which checks that the object status observed generation matches the object metadata generation, along with a few other checks.

#### Worker extension Reconciler
[worker/genericactuator.Reconcile](https://github.com/gardener/gardener/blob/v1.65.2/extensions/pkg/controller/worker/genericactuator/actuator_reconcile.go#L45) is the guy that does the following

`deployMachineControllerManager` # Deploy the machine-controller-manager into the cluster.
➡ `GenerateMachineDeployments` # Generate the desired machine deployments.
➡  `DeployMachineClasses` # Deploy generated machine classes.
➡  `UpdateMachineImagesStatus` # Update image list used by the machines associated with this Worker resource
➡ `deployMachineDeployments` # generate MD configs and deploy them
➡  `waitUntilWantedMachineDeploymentsAvailable` # Wait until all generated MD's re healthy/available.

We *dont* want to wait until all wanted machine deployments are healthy.
